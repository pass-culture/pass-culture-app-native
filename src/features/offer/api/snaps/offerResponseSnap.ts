import { ExpenseDomain, OfferResponse, SubcategoryIdEnum } from 'api/gen'

// humanizedId AHD3A
export const offerResponseSnap: OfferResponse = {
  id: 116656,
  accessibility: {
    audioDisability: true,
    mentalDisability: true,
    motorDisability: false,
    visualDisability: false,
  },
  description:
    'Depuis de nombreuses années, Christine vit sous un pont, isolée de toute famille et amis. Par une nuit comme il n’en existe que dans les contes, un jeune garçon de 8 ans fait irruption devant son abri. Suli ne parle pas français, il est perdu, séparé de sa mère…\nTous les détails du film sur AlloCiné: https://www.allocine.fr/film/fichefilm_gen_cfilm=199293.html',
  expenseDomains: [ExpenseDomain.all],
  isDigital: false,
  isDuo: true,
  isEducational: false,
  name: 'Sous les étoiles de Paris - VF',
  subcategoryId: SubcategoryIdEnum.CINE_PLEIN_AIR,
  isReleased: true,
  isExpired: false,
  isForbiddenToUnderage: false,
  isSoldOut: false,
  stocks: [
    {
      id: 118929,
      beginningDatetime: '2021-01-04T13:30:00',
      price: 500,
      isBookable: true,
      isExpired: false,
      isForbiddenToUnderage: false,
      isSoldOut: false,
    },
    {
      id: 118928,
      beginningDatetime: '2021-01-03T18:00:00',
      price: 500,
      isBookable: true,
      isExpired: false,
      isForbiddenToUnderage: false,
      isSoldOut: false,
    },
  ],
  image: {
    url: 'https://storage.gra.cloud.ovh.net/v1/AUTH_688df1e25bd84a48a3804e7fa8938085/storage-pc-dev/thumbs/products/CHSYS',
    credit: 'Author: photo credit author',
  },
  venue: {
    id: 1664,
    address: '2 RUE LAMENNAIS',
    city: 'PARIS 8',
    offerer: { name: 'PATHE BEAUGRENELLE' },
    name: 'PATHE BEAUGRENELLE',
    postalCode: '75008',
    publicName: undefined,
    coordinates: { latitude: 20, longitude: 2 },
    isPermanent: true,
  },
  withdrawalDetails: 'How to withdraw, https://test.com',
}
