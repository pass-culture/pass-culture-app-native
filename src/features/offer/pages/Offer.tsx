import { t } from '@lingui/macro'
import { StackScreenProps } from '@react-navigation/stack'
import React, { FunctionComponent } from 'react'
import styled from 'styled-components/native'

import { RootStackParamList } from 'features/navigation/RootNavigator'
import { _ } from 'libs/i18n'
import { getSpacing, Spacer, Typo } from 'ui/theme'

import { OfferCategory } from '../atoms/OfferCategory'

type Props = StackScreenProps<RootStackParamList, 'Offer'>

export const Offer: FunctionComponent<Props> = ({ route }: Props) => {
  const { id, algoliaHit } = route.params

  return (
    <Container>
      <Spacer.Flex />
      <Typo.Hero>{_(t`Offer`)}</Typo.Hero>
      <Typo.Caption testID="offerId">{id}</Typo.Caption>
      <Title testID="offerTitle" numberOfLines={3} adjustsFontSizeToFit>
        {algoliaHit?.offer.name}
      </Title>
      <OfferCategory
        category={algoliaHit?.offer.category || null}
        label={algoliaHit?.offer.label}
      />
      <Spacer.Flex />
    </Container>
  )
}

const Container = styled.View({ flex: 1, alignItems: 'center', paddingHorizontal: getSpacing(6) })
const Title = styled(Typo.Title3)({ textAlign: 'center' })
