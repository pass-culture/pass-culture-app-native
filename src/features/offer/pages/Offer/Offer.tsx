import { useFocusEffect, useRoute } from '@react-navigation/native'
import React, { FunctionComponent, useCallback, useEffect } from 'react'
import { NativeScrollEvent, NativeSyntheticEvent } from 'react-native'
import styled from 'styled-components/native'

import { NativeCategoryIdEnumv2, SearchGroupNameEnumv2 } from 'api/gen'
import { StepperOrigin, UseRouteType } from 'features/navigation/RootNavigator/types'
import { useOffer } from 'features/offer/api/useOffer'
import { useSimilarOffers } from 'features/offer/api/useSimilarOffers'
import { BottomBanner } from 'features/offer/components/BottomBanner/BottomBanner'
import { CTAButton } from 'features/offer/components/CTAButton/CTAButton'
import { OfferBody } from 'features/offer/components/OfferBody/OfferBody'
import { OfferHeader } from 'features/offer/components/OfferHeader/OfferHeader'
import { OfferWebHead } from 'features/offer/components/OfferWebHead'
import { PlaylistType } from 'features/offer/enums'
import { getIsFreeDigitalOffer } from 'features/offer/helpers/getIsFreeDigitalOffer/getIsFreeDigitalOffer'
import { getSearchGroupAndNativeCategoryFromSubcategoryId } from 'features/offer/helpers/getSearchGroupAndNativeCategoryFromSubcategoryId/getSearchGroupAndNativeCategoryFromSubcategoryId'
import { useCtaWordingAndAction } from 'features/offer/helpers/useCtaWordingAndAction/useCtaWordingAndAction'
import { analytics, isCloseToBottom } from 'libs/analytics'
import useFunctionOnce from 'libs/hooks/useFunctionOnce'
import { BatchEvent, BatchUser } from 'libs/react-native-batch'
import { useSubcategories } from 'libs/subcategories/useSubcategories'
import { useBookOfferModal } from 'shared/offer/helpers/useBookOfferModal'
import { useOpacityTransition } from 'ui/animations/helpers/useOpacityTransition'
import { getSpacing, Spacer } from 'ui/theme'

const trackEventHasSeenOffer = () => BatchUser.trackEvent(BatchEvent.hasSeenOffer)

const OFFER_NATIVE_CATEGORIES_ELIGIBLE_FOR_SURVEY = [
  NativeCategoryIdEnumv2.SEANCES_DE_CINEMA,
  NativeCategoryIdEnumv2.VISITES_CULTURELLES,
  NativeCategoryIdEnumv2.LIVRES_PAPIER,
  NativeCategoryIdEnumv2.CONCERTS_EVENEMENTS,
]
const trackEventHasSeenOfferForSurvey = () => BatchUser.trackEvent(BatchEvent.hasSeenOfferForSurvey)

const PLAYLIST_HEIGHT = 300

const getPlaylistsHeight = (numberOfPlaylists: number) => {
  return PLAYLIST_HEIGHT * numberOfPlaylists
}

export const Offer: FunctionComponent = () => {
  const route = useRoute<UseRouteType<'Offer'>>()
  const trackEventHasSeenOfferOnce = useFunctionOnce(trackEventHasSeenOffer)
  const trackEventHasSeenOfferForSurveyOnce = useFunctionOnce(trackEventHasSeenOfferForSurvey)
  const offerId = route.params?.id
  const searchId = route.params?.searchId
  const from = route.params?.from

  const trackBookOfferForSurveyOnce = useFunctionOnce(() => {
    BatchUser.trackEvent(BatchEvent.hasSeenBookOfferForSurvey)
  })
  const trackCinemaOfferForSurveyOnce = useFunctionOnce(() => {
    BatchUser.trackEvent(BatchEvent.hasSeenCinemaOfferForSurvey)
  })
  const trackCulturalVisitOfferForSurveyOnce = useFunctionOnce(() => {
    BatchUser.trackEvent(BatchEvent.hasSeenCulturalVisitForSurvey)
  })
  const trackConcertOfferForSurveyOnce = useFunctionOnce(() => {
    BatchUser.trackEvent(BatchEvent.hasSeenConcertForSurvey)
  })

  const { data: offerResponse } = useOffer({ offerId })

  const logConsultWholeOffer = useFunctionOnce(() => {
    if (offerResponse) {
      analytics.logConsultWholeOffer(offerResponse.id)
    }
  })

  const { data: offer } = useOffer({ offerId })
  const { data } = useSubcategories()

  const { searchGroupName, nativeCategory } =
    getSearchGroupAndNativeCategoryFromSubcategoryId(data, offer?.subcategoryId) || {}
  const { similarOffers: sameCategorySimilarOffers, apiRecoParams: apiRecoParamsSameCategory } =
    useSimilarOffers({
      offerId,
      position: offer?.venue.coordinates,
      categoryIncluded: searchGroupName ?? SearchGroupNameEnumv2.NONE,
    })
  const hasSameCategorySimilarOffers = Boolean(sameCategorySimilarOffers?.length)

  const {
    similarOffers: otherCategoriesSimilarOffers,
    apiRecoParams: apiRecoParamsOtherCategories,
  } = useSimilarOffers({
    offerId,
    position: offer?.venue.coordinates,
    categoryExcluded: searchGroupName ?? SearchGroupNameEnumv2.NONE,
  })
  const hasOtherCategoriesSimilarOffers = Boolean(otherCategoriesSimilarOffers?.length)

  const fromOfferId = route.params?.fromOfferId

  const isFreeDigitalOffer = getIsFreeDigitalOffer(offer)

  const shouldTriggerBatchSurveyEvent =
    nativeCategory && OFFER_NATIVE_CATEGORIES_ELIGIBLE_FOR_SURVEY.includes(nativeCategory)

  const logSameCategoryPlaylistVerticalScroll = useFunctionOnce(() => {
    return analytics.logPlaylistVerticalScroll({
      ...apiRecoParamsSameCategory,
      fromOfferId,
      offerId,
      playlistType: PlaylistType.SAME_CATEGORY_SIMILAR_OFFERS,
    })
  })

  const logOtherCategoriesPlaylistVerticalScroll = useFunctionOnce(() => {
    return analytics.logPlaylistVerticalScroll({
      ...apiRecoParamsOtherCategories,
      fromOfferId,
      offerId,
      playlistType: PlaylistType.OTHER_CATEGORIES_SIMILAR_OFFERS,
    })
  })

  const trackBatchEvent = useCallback(() => {
    trackEventHasSeenOfferForSurveyOnce()

    if (nativeCategory === NativeCategoryIdEnumv2.LIVRES_PAPIER) {
      trackBookOfferForSurveyOnce()
    }

    if (nativeCategory === NativeCategoryIdEnumv2.VISITES_CULTURELLES) {
      trackCulturalVisitOfferForSurveyOnce()
    }

    if (nativeCategory === NativeCategoryIdEnumv2.CONCERTS_EVENEMENTS) {
      trackConcertOfferForSurveyOnce()
    }

    if (nativeCategory === NativeCategoryIdEnumv2.SEANCES_DE_CINEMA) {
      trackCinemaOfferForSurveyOnce()
    }
  }, [
    nativeCategory,
    trackBookOfferForSurveyOnce,
    trackCinemaOfferForSurveyOnce,
    trackConcertOfferForSurveyOnce,
    trackCulturalVisitOfferForSurveyOnce,
    trackEventHasSeenOfferForSurveyOnce,
  ])

  const handleLogPlaylistVerticalScroll = useCallback(
    (nativeEvent: NativeScrollEvent) => {
      // The log event is triggered when the similar offer playlist is visible
      const hasTwoSimilarOffersPlaylist =
        hasSameCategorySimilarOffers && hasOtherCategoriesSimilarOffers

      if (
        isCloseToBottom({
          ...nativeEvent,
          padding: getPlaylistsHeight(2),
        }) &&
        hasTwoSimilarOffersPlaylist
      ) {
        logSameCategoryPlaylistVerticalScroll()
      }

      if (
        isCloseToBottom({
          ...nativeEvent,
          padding: getPlaylistsHeight(1),
        })
      ) {
        if (hasTwoSimilarOffersPlaylist || hasOtherCategoriesSimilarOffers) {
          logOtherCategoriesPlaylistVerticalScroll()
        } else if (!hasTwoSimilarOffersPlaylist && hasSameCategorySimilarOffers) {
          logSameCategoryPlaylistVerticalScroll()
        }
      }
    },
    [
      hasOtherCategoriesSimilarOffers,
      hasSameCategorySimilarOffers,
      logOtherCategoriesPlaylistVerticalScroll,
      logSameCategoryPlaylistVerticalScroll,
    ]
  )

  const scrollEventListener = useCallback(
    ({ nativeEvent }: NativeSyntheticEvent<NativeScrollEvent>) => {
      if (isCloseToBottom(nativeEvent)) {
        logConsultWholeOffer()
        if (shouldTriggerBatchSurveyEvent) {
          trackBatchEvent()
        }
      }
      handleLogPlaylistVerticalScroll(nativeEvent)
    },
    [
      handleLogPlaylistVerticalScroll,
      logConsultWholeOffer,
      shouldTriggerBatchSurveyEvent,
      trackBatchEvent,
    ]
  )

  const { headerTransition, onScroll } = useOpacityTransition({
    listener: scrollEventListener,
  })

  const {
    wording,
    onPress: onPressCTA,
    navigateTo,
    externalNav,
    modalToDisplay,
    isEndedUsedBooking,
    bottomBannerText,
    isDisabled,
  } = useCtaWordingAndAction({ offerId, from, searchId }) ?? {}

  const { OfferModal: CTAOfferModal, showModal: showOfferModal } = useBookOfferModal({
    modalToDisplay,
    offerId,
    isEndedUsedBooking,
    from: StepperOrigin.OFFER,
  })

  useFocusEffect(
    useCallback(() => {
      trackEventHasSeenOfferOnce()
      if (route.params.openModalOnNavigation) {
        showOfferModal()
      }
    }, [trackEventHasSeenOfferOnce, route.params.openModalOnNavigation, showOfferModal])
  )

  useEffect(() => {
    let timeoutId: NodeJS.Timeout
    if (shouldTriggerBatchSurveyEvent) {
      timeoutId = setTimeout(() => {
        trackBatchEvent()
      }, 5000)
    }

    return () => clearTimeout(timeoutId)
  }, [shouldTriggerBatchSurveyEvent, trackBatchEvent])

  const onPress = () => {
    onPressCTA?.()
    showOfferModal()
  }

  if (!offerResponse) return null

  return (
    <Container>
      <OfferWebHead offer={offerResponse} />
      <OfferBody
        offerId={offerId}
        onScroll={onScroll}
        sameCategorySimilarOffers={sameCategorySimilarOffers}
        apiRecoParamsSameCategory={apiRecoParamsSameCategory}
        otherCategoriesSimilarOffers={otherCategoriesSimilarOffers}
        apiRecoParamsOtherCategories={apiRecoParamsOtherCategories}
      />
      {/* OfferHeader is called after Body to implement the BlurView for iOS */}
      <OfferHeader
        title={offerResponse.name}
        headerTransition={headerTransition}
        offerId={offerResponse.id}
        searchId={searchId}
      />
      {!!wording && (
        <React.Fragment>
          <CallToActionContainer testID="CTA-button">
            <CTAButton
              wording={wording}
              onPress={onPress}
              navigateTo={navigateTo}
              externalNav={externalNav}
              isDisabled={isDisabled}
              isFreeDigitalOffer={isFreeDigitalOffer}
            />
            <Spacer.Column numberOfSpaces={bottomBannerText ? 4.5 : 6} />
          </CallToActionContainer>
          {bottomBannerText ? <BottomBanner text={bottomBannerText} /> : <Spacer.BottomScreen />}
        </React.Fragment>
      )}

      {CTAOfferModal}
    </Container>
  )
}

const Container = styled.View(({ theme }) => ({
  flex: 1,
  backgroundColor: theme.colors.white,
}))

const CallToActionContainer = styled.View({
  marginHorizontal: getSpacing(6),
})
