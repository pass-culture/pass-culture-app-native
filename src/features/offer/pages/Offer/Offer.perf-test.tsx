import React from 'react'
// eslint-disable-next-line import/no-extraneous-dependencies
import { measurePerformance } from 'reassure'
import { ThemeProvider } from 'styled-components/native'

import { OfferResponse } from 'api/gen'
import { offerResponseSnap } from 'features/offer/fixtures/offerResponse'
import * as InstalledAppsCheck from 'features/offer/helpers/checkInstalledApps/checkInstalledApps'
import { Offer } from 'features/offer/pages/Offer/Offer'
import { placeholderData } from 'libs/subcategories/placeholderData'
import { computedTheme } from 'tests/computedTheme'
import { reactQueryProviderHOC } from 'tests/reactQueryProviderHOC'
import { waitFor } from 'tests/utils'
import { Network } from 'ui/components/ShareMessagingApp'

const mockedOffer: Partial<OfferResponse> | undefined = offerResponseSnap
jest.mock('features/offer/api/useOffer', () => ({
  useOffer: () => ({
    data: mockedOffer,
  }),
}))

const mockSubcategories = placeholderData.subcategories
jest.mock('libs/subcategories/useSubcategories', () => ({
  useSubcategories: () => ({
    data: {
      subcategories: mockSubcategories,
    },
  }),
}))

const mockCheckInstalledApps = jest.spyOn(InstalledAppsCheck, 'checkInstalledApps') as jest.Mock
mockCheckInstalledApps.mockResolvedValue({
  [Network.snapchat]: true,
})

test('Simple test', async () => {
  await waitFor(async () => {
    await measurePerformance(<Offer />, {
      wrapper: (children) =>
        // eslint-disable-next-line local-rules/no-react-query-provider-hoc
        reactQueryProviderHOC(<ThemeProvider theme={computedTheme}>{children}</ThemeProvider>),
    })
  })
})
