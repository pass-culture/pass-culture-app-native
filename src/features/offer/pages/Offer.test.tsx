import { NavigationContainer } from '@react-navigation/native'
import { act, fireEvent, render } from '@testing-library/react-native'
import { rest } from 'msw/'
import React from 'react'
import { UseQueryResult } from 'react-query'
import { ReactTestInstance } from 'react-test-renderer'
import waitForExpect from 'wait-for-expect'

import { OfferResponse, UserProfileResponse } from 'api/gen'
import { RootStack } from 'features/navigation/RootNavigator'
import { logConsultAccessibility, logConsultWithdrawal, logConsultWholeOffer } from 'libs/analytics'
import { env } from 'libs/environment'
import { reactQueryProviderHOC } from 'tests/reactQueryProviderHOC'
import { server } from 'tests/server'
import { flushAllPromises } from 'tests/utils'

import { offerResponseSnap } from '../api/snaps/offerResponseSnap'
import { dehumanizeId } from '../services/dehumanizeId'

import { Offer } from './Offer'

jest.mock('libs/geolocation')

jest.mock('@react-navigation/native', () => jest.requireActual('@react-navigation/native'))

jest.mock('features/auth/AuthContext', () => ({
  useAuthContext: jest.fn(() => ({ isLoggedIn: true })),
}))

jest.mock('features/home/api', () => ({
  useUserProfileInfo: jest.fn(
    () =>
      ({
        isLoading: false,
        data: { email: 'email2@domain.ext', firstName: 'Jean', isBeneficiary: true },
      } as UseQueryResult<UserProfileResponse>)
  ),
}))

describe('<Offer />', () => {
  it('should match snapshot for physical offer', async () => {
    const { toJSON } = await renderOfferPage({ isDigital: false })
    expect(toJSON()).toMatchSnapshot()
  })

  it('should match snapshot for digital offer', async () => {
    const { toJSON } = await renderOfferPage({ isDigital: true, isDuo: false })
    expect(toJSON()).toMatchSnapshot()
  })
  it('should show withdrawalDetails', async () => {
    let wrapper = await renderOfferPage({ withdrawalDetails: 'How to withdraw' })
    expect(wrapper.queryByText('Modalités de retrait')).toBeTruthy()

    wrapper = await renderOfferPage({ withdrawalDetails: undefined })
    expect(wrapper.queryByText('Modalités de retrait')).toBeFalsy()
  })
  it('animates on scroll', async () => {
    const { getByTestId } = await renderOfferPage()
    expect(getByTestId('offerHeaderName').props.style.opacity).toBe(0)
    const scrollContainer = getByTestId('offer-container')
    await act(async () => await fireEvent.scroll(scrollContainer, scrollEvent))
    expect(getByTestId('offerHeaderName').props.style.opacity).toBe(1)
  })

  it('should not show distance and go to button', async () => {
    const wrapper = await renderOfferPage({
      venue: {
        id: 1664,
        address: '2 RUE LAMENNAIS',
        city: 'PARIS 8',
        offerer: { name: 'PATHE BEAUGRENELLE' },
        name: 'PATHE BEAUGRENELLE',
        postalCode: '75008',
        publicName: undefined,
        coordinates: {},
      },
    })
    expect(wrapper.queryByText("Voir l'itinéraire")).toBeFalsy()
    expect(wrapper.queryByText('Distance')).toBeFalsy()
  })

  describe('Analytics', () => {
    beforeAll(() => jest.useFakeTimers())
    const trigger = (component: ReactTestInstance) => {
      act(() => {
        fireEvent.press(component)
        jest.advanceTimersByTime(300)
      })
    }

    it('should log ConsultAccesibilityModalities once when opening accessibility modalities', async () => {
      const { getByText } = await renderOfferPage()

      trigger(getByText('Accessibilité'))
      expect(logConsultAccessibility).toHaveBeenCalledTimes(1)
      expect(logConsultAccessibility).toHaveBeenCalledWith(offerId)

      trigger(getByText('Accessibilité'))
      trigger(getByText('Accessibilité'))
      expect(logConsultAccessibility).toHaveBeenCalledTimes(1)
    })
    it('should log ConsultWithdrawalModalities once when opening accessibility modalities', async () => {
      const { getByText } = await renderOfferPage()

      trigger(getByText('Modalités de retrait'))
      expect(logConsultWithdrawal).toHaveBeenCalledTimes(1)
      expect(logConsultWithdrawal).toHaveBeenCalledWith(offerId)

      trigger(getByText('Modalités de retrait'))
      trigger(getByText('Modalités de retrait'))
      expect(logConsultWithdrawal).toHaveBeenCalledTimes(1)
    })

    const nativeEventMiddle = {
      layoutMeasurement: { height: 1000 },
      contentOffset: { y: 400 }, // how far did we scroll
      contentSize: { height: 1600 },
    }
    const nativeEventBottom = {
      layoutMeasurement: { height: 1000 },
      contentOffset: { y: 900 },
      contentSize: { height: 1600 },
    }

    it('should trigger logEvent "ConsultAllOffer" when reaching the end', async () => {
      const offerPage = await renderOfferPage()
      const scrollView = offerPage.getByTestId('offer-container')

      await act(async () => {
        await scrollView.props.onScroll({ nativeEvent: nativeEventMiddle })
        // await flushAllPromises()
      })
      expect(logConsultWholeOffer).not.toHaveBeenCalled()

      await act(async () => {
        await scrollView.props.onScroll({ nativeEvent: nativeEventBottom })
      })

      expect(logConsultWholeOffer).toHaveBeenCalledWith(offerId)
    })

    it('should trigger logEvent "ConsultAllOffer" only once', async () => {
      const offerPage = await renderOfferPage()
      const scrollView = offerPage.getByTestId('offer-container')
      await act(async () => {
        // 1st scroll to bottom => trigger
        await scrollView.props.onScroll({ nativeEvent: nativeEventBottom })
      })
      expect(logConsultWholeOffer).toHaveBeenCalledWith(offerId)

      // @ts-ignore: logConsultWholeOffer is the mock function but is seen as the real function
      logConsultWholeOffer.mockClear()

      await act(async () => {
        // 2nd scroll to bottom => NOT trigger
        await scrollView.props.onScroll({ nativeEvent: nativeEventMiddle })
        await scrollView.props.onScroll({ nativeEvent: nativeEventBottom })
      })

      expect(logConsultWholeOffer).not.toHaveBeenCalled()
    })
  })
})

const scrollEvent = {
  nativeEvent: {
    contentOffset: { y: 200 },
    layoutMeasurement: { height: 1000 },
    contentSize: { height: 1600 },
  },
}

const humanizedOfferId = 'AHD3A'
const offerId = dehumanizeId(humanizedOfferId)

async function renderOfferPage(extraOffer?: Partial<Omit<OfferResponse, 'id'>>) {
  server.use(
    rest.get<OfferResponse>(env.API_BASE_URL + `/native/v1/offer/${offerId}`, (req, res, ctx) =>
      res.once(ctx.status(200), ctx.json({ ...offerResponseSnap, ...extraOffer }))
    )
  )

  const wrapper = render(
    reactQueryProviderHOC(
      <NavigationContainer>
        <RootStack.Navigator initialRouteName="Offer">
          <RootStack.Screen
            name="Offer"
            component={Offer}
            initialParams={{ id: humanizedOfferId }}
          />
        </RootStack.Navigator>
      </NavigationContainer>
    )
  )

  await act(async () => {
    await flushAllPromises()
  })

  await waitForExpect(() => {
    expect(wrapper.queryByTestId('offer-container')).toBeTruthy()
  })

  return wrapper
}
