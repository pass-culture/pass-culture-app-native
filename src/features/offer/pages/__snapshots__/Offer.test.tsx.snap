// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Offer /> should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "zIndex": 1,
          },
          false,
        ]
      }
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        onLayout={[Function]}
        pointerEvents="box-none"
        style={null}
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 0,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(255, 255, 255)",
                "borderBottomColor": "rgb(216, 216, 216)",
                "flex": 1,
                "shadowColor": "rgb(216, 216, 216)",
                "shadowOffset": Object {
                  "height": 0.5,
                  "width": 0,
                },
                "shadowOpacity": 0.85,
                "shadowRadius": 0,
              }
            }
          />
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "height": 64,
              "maxHeight": undefined,
              "minHeight": undefined,
              "opacity": undefined,
              "transform": undefined,
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "height": 20,
              }
            }
          />
          <View
            pointerEvents="box-none"
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "marginHorizontal": 16,
                  "opacity": 1,
                }
              }
            >
              <Text
                accessibilityRole="header"
                aria-level="1"
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Object {
                    "color": "rgb(28, 28, 30)",
                    "fontSize": 17,
                    "fontWeight": "600",
                  }
                }
              >
                Offer
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "opacity": 1,
            }
          }
        />
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onGestureEnd={[Function]}
          onOpen={[Function]}
          onTransitionStart={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              collapsable={false}
              forwardedRef={[Function]}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              style={
                Object {
                  "flex": 1,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "rgb(242, 242, 242)",
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": -1,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 5,
                    "top": 0,
                    "width": 3,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "alignItems": "center",
                            "flexBasis": 0,
                            "flexGrow": 1,
                            "flexShrink": 1,
                            "paddingHorizontal": 24,
                          },
                        ]
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "flexBasis": 0,
                              "flexGrow": 1,
                              "flexShrink": 1,
                            },
                          ]
                        }
                      />
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#151515",
                              "fontFamily": "Montserrat-Medium",
                              "fontSize": 38,
                              "lineHeight": 48,
                            },
                          ]
                        }
                      >
                        Offer
                      </Text>
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#151515",
                              "fontFamily": "Montserrat-SemiBold",
                              "fontSize": 12,
                              "lineHeight": 16,
                            },
                          ]
                        }
                        testID="offerId"
                      >
                        ABCDE
                      </Text>
                      <Text
                        adjustsFontSizeToFit={true}
                        numberOfLines={3}
                        style={
                          Array [
                            Object {
                              "color": "#151515",
                              "fontFamily": "Montserrat-Bold",
                              "fontSize": 20,
                              "lineHeight": 24,
                              "textAlign": "center",
                            },
                          ]
                        }
                        testID="offerTitle"
                      >
                        Mensch ! OÃ¹ sont les Hommes ?
                      </Text>
                      <View
                        style={
                          Array [
                            Object {
                              "alignItems": "center",
                              "flexBasis": 0,
                              "flexGrow": 1,
                              "flexShrink": 1,
                            },
                          ]
                        }
                      >
                        <RNSVGSvgView
                          align="xMidYMid"
                          bbHeight={48}
                          bbWidth={48}
                          fill="#626262"
                          focusable={false}
                          height={48}
                          meetOrSlice={0}
                          minX={20}
                          minY={20}
                          style={
                            Array [
                              Object {
                                "backgroundColor": "transparent",
                                "borderWidth": 0,
                              },
                              Object {
                                "flex": 0,
                                "height": 48,
                                "width": 48,
                              },
                            ]
                          }
                          vbHeight={142}
                          vbWidth={142}
                          width={48}
                        >
                          <RNSVGGroup
                            fill={4284637794}
                            propList={
                              Array [
                                "fill",
                              ]
                            }
                          >
                            <RNSVGPath
                              d="M66.72 123.64c8.13 0 14.72 6.59 14.72 14.72s-6.59 14.72-14.72 14.72S52 146.49 52 138.36s6.59-14.72 14.72-14.72zm0 6c-4.816 0-8.72 3.904-8.72 8.72 0 4.816 3.904 8.72 8.72 8.72 4.816 0 8.72-3.904 8.72-8.72 0-4.816-3.904-8.72-8.72-8.72zm70.225-89.705c2.494 1.918 4.038 4.803 4.266 7.921l.029.784v65.228c.027.36.04.725.04 1.092 0 8.13-6.59 14.72-14.72 14.72s-14.72-6.59-14.72-14.72 6.59-14.72 14.72-14.72c3.246 0 6.246 1.05 8.68 2.83V48.646c-.003-1.55-.724-3.01-1.952-3.954-1.075-.827-2.432-1.176-3.78-.98l-.575.117L84.98 54.897c-2.014.542-3.461 2.27-3.673 4.307l-.027.516V112c0 1.657-1.343 3-3 3-1.519 0-2.774-1.129-2.973-2.593L75.28 112V59.718c.003-4.734 3.03-8.907 7.505-10.42l.682-.207 43.968-11.072c3.296-.869 6.808-.161 9.51 1.916zM126.56 106.24c-4.816 0-8.72 3.904-8.72 8.72 0 4.816 3.904 8.72 8.72 8.72 4.676 0 8.492-3.68 8.71-8.302l-.003-.01-.027-.408v-.84c-.422-4.422-4.147-7.88-8.68-7.88z"
                              matrix={
                                Array [
                                  1,
                                  0,
                                  0,
                                  1,
                                  0,
                                  0,
                                ]
                              }
                            />
                          </RNSVGGroup>
                        </RNSVGSvgView>
                        <Text
                          style={
                            Array [
                              Object {
                                "color": "#151515",
                                "fontFamily": "Montserrat-SemiBold",
                                "fontSize": 12,
                                "lineHeight": 16,
                              },
                            ]
                          }
                        >
                          Abonnements concerts
                        </Text>
                      </View>
                      <View
                        style={
                          Array [
                            Object {
                              "flexBasis": 0,
                              "flexGrow": 1,
                              "flexShrink": 1,
                            },
                          ]
                        }
                      />
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
